clang/lib/CIR/CodeGen/CIRGenExpr.cpp

```diff
@@ -1178,11 +1178,11 @@ static void pushTemporaryCleanup(CIRGenFunction &cgf,
                                         ->getBaseElementTypeUnsafe()
                                         ->getAs<clang::RecordType>()) {
     // Get the destructor for the reference temporary.
-    auto *classDecl =
-        cast<CXXRecordDecl>(rt->getOriginalDecl()->getDefinitionOrSelf());
-    if (!classDecl->hasTrivialDestructor())
-      referenceTemporaryDtor =
-          classDecl->getDefinitionOrSelf()->getDestructor();
+    if (const auto *classDecl = dyn_cast<CXXRecordDecl>(
```

Comment:
Since this was a `cast` before, it would have asserted if the return value was null. Are there cases where this returns null?

Reply:
Yes, this asserted in C mode where the record was not a C++ class. Note that OGCG is also dyn_casting here.

Reply:
`test3()` contains a `MaterializeTemporaryExpr` in C: https://godbolt.org/z/b7WY5dfYW